### ISSUE TITLE
Migrating working dockerfile with centos:08/ibm_db/zOS license to rhel8: on cirrus  -- License fails

### ISSUE THREAD

**Issue by debrajo** [ibmdb/node-ibm_db#690]
I have a working dockerfile using  centos:08 / ibm_db / zOS license that runs in a cirrus environment.  I'm trying to convert it to use  ubi8/nodejs-10.  The new Dockerfile connects successfully on my local Windows Docker.  But when deployed to cirrus, I receive a license file error connecting to zOS.   I've recreated the error under a simpler docker to test and no combination of permissions allows it to be successful.  Cirrus team hasn't been too helpful. 

Can you give any other suggestions? 

[Dockerfile.txt](https://github.com/ibmdb/node-ibm_db/files/5170347/Dockerfile.txt)
[docker_db_execute.txt](https://github.com/ibmdb/node-ibm_db/files/5170350/docker_db_execute.txt)
[code.txt](https://github.com/ibmdb/node-ibm_db/files/5170379/code.txt)

**Comment by bimalkjha** [comment]
@debrajo There are two solutions as below. You can choose either:
1. Use `db2consv_zs.lic` of version 11.5 as we have updated the default clidriver for Open source drivers to Db2 11.5.4.0 recently.
2. Update your Dockerfile.txt to download `https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/linuxx64/linuxx64_odbc_cli.tar.gz` instead of `https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/linuxx64_odbc_cli.tar.gz`. It is the older one and will work with existing license of Db2 v11.1.
Thanks.

**Comment by debrajo** [comment]
I've already upgraded to V11.5 license.   Swapped to the  odbc_cli/linuxx64/linuxx64_odbc_cli.tar.gz   copy and that did not fix the problem 
`error: '[node-ibm_db] SQL_ERROR',
2020-09-03T21:13:30.699Z | daj-db-585dff55cd-dhvds |   sqlcode: -1598,
2020-09-03T21:13:30.699Z | daj-db-585dff55cd-dhvds |   message:
2020-09-03T21:13:35.352Z | daj-db-585dff55cd-dhvds | testQuery.test connection error
2020-09-03T21:13:35.353Z | daj-db-585dff55cd-dhvds |    '[IBM][CLI Driver] SQL1598N  An attempt to connect to the database server failed because of a licensing problem.  SQLSTATE=42968\n',
2020-09-03T21:13:35.353Z | daj-db-585dff55cd-dhvds |   state: '42968' }`

**Comment by debrajo** [comment]
Do you have an older, pre-Aug 26  tar.gz file I can try.   Somehow I'm remembering that I got it work once in my cirrus dev, then promoted to uat ..  and it failed, repromoted to dev .. and dev / uat failing ..   The switch to ubi8:node10 was right before/after that timeframe.

**Comment by bimalkjha** [comment]
@debrajo `https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/linuxx64/linuxx64_odbc_cli.tar.gz` is the pre-Aug26 tar.gz which requires v11.1 license. 
If you have v11.5 license, then latest ibm_db should work without any change. I would suggest to run below commands and share the result.
```
npm install ibm_db@2.7.1
copy v11.5 db2consv_zs.lic to clidriver/license directory.
source installer/setenv.sh
db2cli validate -database "dbname:hostname:port" -user dbuser -passwd dbpass -connect -displaylic
```
Thanks.

**Comment by debrajo** [comment]
Adding the above to my run.sh .. also tried to add tracing.  
To get the license to work successfully locally, I had to swap back to the odbc_cli/linuxx64_odbc_cli.tar.gz.  Using odbc_cli/linuxx64/linuxx64_odbc_cli.tar.gz locally, I got the license error. 

Locally, I received the following success 
[ibm_db_localhost_validate.txt](https://github.com/ibmdb/node-ibm_db/files/5174859/ibm_db_localhost_validate.txt)

Deployed same to cirrus.  it fails,  here's the result  (and no trace files, not enough space on container) 
[ibm_db_cirrus_validate.txt](https://github.com/ibmdb/node-ibm_db/files/5174866/ibm_db_cirrus_validate.txt)

Since it can't find the license, this is all about visibility of license to the container.   I'm going to continue to work with cirrus team on this.  

Any other suggestions?

**Comment by debrajo** [comment]
I updated by  run.sh  to perform better ls -al and this is the result:
` ls of /ibmdriver/clidriver/license
 -r-xr-xr-x.  1 default root  1854 May  7 14:01 odbc_REDIST.txt
 -r-xr-xr-x.  1 default root 23445 Oct  9  2019 odbc_notices.txt
 ls -al ./node_modules/ibm_db/installer/clidriver/license
 total 36
 drwxr-xr-x.  3 default root    88 Sep  4 13:23 .
 drwxr-xr-x. 13 default root   152 Aug  6 21:48 ..
 drwxr-xr-x.  2 default root  4096 Jun 16 19:53 UNIX
 -rwxrwxr-x.  1 default root   915 Sep  4 13:23 db2consv_zs.lic
 drwxrwxr-x. 2 default root  29 Sep  4 13:24 .
 drwxr-xr-x. 3 default root  21 Sep  4 13:24 ..
 -rwxrwxr-x. 1 default root 915 Sep  4 13:24 db2consv_zs.lic
 total 4
 start validate
......
 ===============================================================================
 Connection attempt for database "EUBADB2A:meuba.vipa.uk.ibm.com:446":
 ===============================================================================
 [FAILED]: [IBM][CLI Driver] SQL1598N  An attempt to connect to the database server failed because of a licensing problem.  SQLSTATE=42968
 ===============================================================================
 Valid License Found            : No
 License location               : N/A
 Server routine type attempted  : SP and UDF
 DB2 connect license version    : N/A
 ===============================================================================
 Error: The validation operation failed.`

Since both ls -al  found license files,  shouldn't the  ibm_db  find the license?

**Comment by debrajo** [comment]
sorry.. blast paste ..  there's soemthing in the log stream that made it uppercase !

**Comment by bimalkjha** [comment]
@debrajo If you check the `ibm_db_cirus_cant_find_license.txt` file, the `ls of /ibmdriver/clidriver/license` do not show db2consv_zs.lic file. Hence the error.
Also, why two clidriver directory your have? `/ibmdriver/clidriver/license` and `node_modules/ibm_db/installer/clidriver/license` both? 
Why ` ls of node_modules/ibm_db/installer/clidriver/license` do not shows the files listed by ``ls of /ibmdriver/clidriver/license`?

**Comment by bimalkjha** [comment]
Contents of all three files shared by you says that, for local setup, you have downloaded clidriver of v11.1 and untared it as `/ibmdriver/clidriver/license`. Then IBM_DB_HOME is environment variable is set to /ibmdriver/clidriver, so it is working fine.
But, for cirrus, the environment variable IBM_DB_HOME is not set and hence ibm_db downloads latest clidriver as `node_modules/ibm_db/installer/clidriver`. If IBM_DB_HOME is set in cirrus environment before `npm install ibm_db`, `node_modules/ibm_db/installer/clidriver` will not get downloaded. Also, note that `node_modules/ibm_db/installer/clidriver` requires v11.5 db2consv_zs.lic file. Thanks.

**Comment by debrajo** [comment]
The Dockerfile being run locally and in cirrus are identical.  I perform all my work locally, run docker build locally, makes sure it works, then promote to github.  Cirrus pulls github and runs Dockerfile.    Comparing the two logs, they are identical.  Both logs have 'IBM_DB_HOME environment variable have already been set to -> /ibmdriver/clidriver' .  And I previously checked, there is nothing in  node_modules/ibm_db  

Based on your comments, I switched up the dockerfile to not download linux64_odbc.cli.tar.gz to /ibmdriver and let npm install download and install.    It successfully connects with license to MVS.  This tells me it's really about permissions for the /ibmdriver folder and license ..   And after much trial and error, I got it to work.   

For everyone else's reference.  Here's the working dockerfile for cirrus / linux 
[ibm_db_cirrus_working_dockerffile.txt](https://github.com/ibmdb/node-ibm_db/files/5189011/ibm_db_cirrus_working_dockerffile.txt)

I do thank you for your help.  Sometimes, it takes another persons perspective / ideas to rethink the problem.
